lib/integrations/azureoidc/enterprise_app.go

```diff
@@ -0,0 +1,168 @@
+	spID := *sp.GetId()
+
+	msGraphResourceID, err := getMSGraphResourceID(ctx, graphClient)
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err, "failed to get MS Graph API resource ID")
+	}
+
+	msGraphResourceUUID := uuid.MustParse(msGraphResourceID)
+
+	for _, appRoleID := range appRoles {
+		assignment := models.NewAppRoleAssignment()
+		spUUID := uuid.MustParse(spID)
+		assignment.SetPrincipalId(&spUUID)
+
+		assignment.SetResourceId(&msGraphResourceUUID)
+
+		appRoleUUID := uuid.MustParse(appRoleID)
+		assignment.SetAppRoleId(&appRoleUUID)
+		_, err := graphClient.ServicePrincipals().
+			ByServicePrincipalId(spID).
+			AppRoleAssignments().
+			Post(ctx, assignment, nil)
+		if err != nil {
+			return appID, tenantID, trace.Wrap(err, "failed to assign app role %s", appRoleID)
+		}
+	}
+
+	if err := createFederatedAuthCredential(ctx, graphClient, *app.GetId(), proxyPublicAddr); err != nil {
+		return appID, tenantID, trace.Wrap(err, "failed to create an OIDC federated auth credential")
+	}
+
+	acsURL, err := url.Parse(proxyPublicAddr)
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err, "failed to parse proxy public address")
+	}
+	acsURL.Path = path.Join("/v1/webapi/saml/acs", authConnectorName)
+	if err := setupSSO(ctx, graphClient, *app.GetId(), spID, acsURL.String()); err != nil {
+		return appID, tenantID, trace.Wrap(err, "failed to set up SSO for the enterprise app")
+	}
+
+	return appID, tenantID, nil
+}
+
+// createFederatedAuthCredential creates a new federated (OIDC) auth credential for the given Entra application.
+func createFederatedAuthCredential(ctx context.Context, graphClient *msgraphsdk.GraphServiceClient, appObjectID string, proxyPublicAddr string) error {
+	credential := models.NewFederatedIdentityCredential()
+	name := "teleport-oidc"
+	audiences := []string{azureDefaultJWTAudience}
+	subject := azureSubject
+	credential.SetName(&name)
+	credential.SetIssuer(&proxyPublicAddr)
+	credential.SetAudiences(audiences)
+	credential.SetSubject(&subject)
+
+	// ByApplicationID here means the object ID,
+	// i.e. app.GetId(), not app.GetAppId().
+	_, err := graphClient.Applications().ByApplicationId(appObjectID).
+		FederatedIdentityCredentials().Post(ctx, credential, nil)
+
+	return trace.Wrap(err)
+
+}
+
+// getMSGraphResourceID gets the resource ID for the Microsoft Graph app in the Entra directory.
+func getMSGraphResourceID(ctx context.Context, graphClient *msgraphsdk.GraphServiceClient) (string, error) {
+	requestFilter := "displayName eq 'Microsoft Graph'"
+
+	requestParameters := &serviceprincipals.ServicePrincipalsRequestBuilderGetQueryParameters{
+		Filter: &requestFilter,
+	}
+	configuration := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
+		QueryParameters: requestParameters,
+	}
+	spResponse, err := graphClient.ServicePrincipals().Get(ctx, configuration)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	spList := spResponse.GetValue()
+	if len(spList) < 1 {
```

Comment:
To make sure only one is returned?
```suggestion
	if len(spList) != 1 {
```